### PR TITLE
Add bitwise operations validator pass

### PR DESCRIPTION
**Pull request by atgoo**
The pass checks correctness of operand types of all bitwise instructions
(opcode range from SpvOpShiftRightLogical to SpvOpBitCount).

### CODE REVIEW

**Comment by atgoo** [comment]
vk-gl-cts master test successful.

File: source/validate_bitwise.cpp [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+#include "val/instruction.h"
+#include "val/validation_state.h"
+
+namespace libspirv {
+
+namespace {
+
+// Returns operand word for given instruction and operand index.
+// The operand is expected to only have one word.
+inline uint32_t GetOperandWord(const spv_parsed_instruction_t* inst,
```
Comment (dneto0) [inline source/validate_bitwise.cpp]: I've seen this a few times.

A future CL could refactor this into a common spot in source/instruction.cpp.

However, the better change is a refactor of the internal representation of SPIR-V instructions, commonizing functionality between the optimizer and the validator.

So I'm happy to wait.  This is fine.

File: source/validate_bitwise.cpp [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+    case SpvOpBitCount: {
+      if (!_.IsIntScalarType(result_type) &&
+          !_.IsIntVectorType(result_type))
+        return _.diag(SPV_ERROR_INVALID_DATA)
+            << "Expected int scalar or vector type as Result Type: "
+            << spvOpcodeString(opcode);
+
+      const uint32_t base_type = GetOperandTypeId(_, inst, 2);
+        if (!base_type ||
```
Comment (dneto0) [inline source/validate_bitwise.cpp]: Indentation is weird.

File: source/validate_bitwise.cpp [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+              << "Expected Base Type to be int scalar or vector: "
+              << spvOpcodeString(opcode);
+
+      const uint32_t base_dimension = _.GetDimension(base_type);
+      const uint32_t result_dimension = _.GetDimension(result_type);
+
+      if (base_dimension != result_dimension)
+        return _.diag(SPV_ERROR_INVALID_DATA)
+            << "Expected Base dimension to be equal to Result Type dimension: "
+            << spvOpcodeString(opcode);
```
Comment (dneto0) [inline source/validate_bitwise.cpp]: Also for Result Type:  "The components must be wide enough to hold the unsigned Width of Base as an unsigned value. That is, no sign bit is needed or counted when checking for a wide enough result width."

So you should be doing a width check here too.

**Comment by dneto0** [comment]
Rebased and pushd into master as 64d5e5214f3

File: source/validate_bitwise.cpp (hunk 1 above)
Comment (atgoo) [reply source/validate_bitwise.cpp]: But this is not Instruction, this is spv_parsed_instruction_t. There seems to be no good place for it.

**Pull request closed without merging**